sql/src/main/java/io/crate/execution/engine/indexing/ShardingUpsertExecutor.java

```diff
@@ -333,4 +337,25 @@ private void maybeSetInterrupt(@Nullable Exception failure) {
         }
     }
 
+    private static class HeapThresholdChecker implements Predicate<ShardedRequests<?, ?>> {
+        private final Runtime rt;
+        private final AtomicLong threshold;
+
+        HeapThresholdChecker(int freeHeapPercent) {
+            if (freeHeapPercent <= 0 || freeHeapPercent >= 100) {
+                throw new IllegalArgumentException(String.format("value must be between 0..100: %d", freeHeapPercent));
+            }
+            rt = Runtime.getRuntime();
+            threshold = new AtomicLong(rt.maxMemory() * freeHeapPercent / 100);
+        }
+
+        boolean enoughFreeMemory() {
+            return rt.freeMemory() >= threshold.get();
```

Comment:
What's the reasoning to go with `Runtime.freeMemory`   instead of utilizing the CircuitBreaker facilities?

Reply:
I had an attempt in that direction. 

The goal was to create a new "entry" in `CrateCircuitBreakerService`, (also in `CircuitBreakers` to produce JMX-accessible stats). But then i found myself needing to inject the CCBS from the top, i.e.`ShardCollectSource` into `ShardCollectorProviderFactory` ... `Blob/LuceneShardCollectorProvider` ... `ShardCollectorProvider`, `ProjectionToProjectorVisitor`, `Column/IndexWriterProje tor` to finally get access to it from `ShardingUpsertExecutor`, where I would have used the CCBS to create a `RamAccountingContext`. 
Then the plan was to use this RAC to add bytes on per `ShardUpsertRequest` and substract them on per `ShardResponse`, as received by `ShardResponseActionListener`. But I had no link between these request/responses and the amount of bytes they served. 
I would also have had to review a few tests.

Time was running out and I wanted to have some to show, at least to see if I was in the right track, and if not, to get back into it asap. My rationale is that I might estimate heap utilisation based on byte accounting, or I may look at the heap and check how free it actually is. There might be other minor GCs happening under the hood, so this measurement could be more up to date. I also thought in changing the implementation of the test method, to account for heap filling up, and heap flushing, that is, the threshold is crossed, but coming from the correct direction.

How should I proceed?. Thank you in advance.

Reply:
My concern here is that we end up running into deadlocks, where several concurrent operations each use up a bit until the heap limit is reached, then all of them pause and never resume again, as they're the components using up the memory themselves. 

Seeing how some tests are now running in a timeout, this might already be the case.

Reply:
As an intermediate improvement we could just use this for the partitioning, but don't pause at all based on memory usage.
Then we do some testing to see how it behaves and if necessary we go back to the drawing board.

Reply:
It all makes sense. I was aware of a different method for checking available heap space, which involved the system's memory mxbean. I have created a benchmark to gain insight. The runtime method is still better performing:

```
Benchmark                                                Mode  Cnt    Score      Error  Units
 * CheckingFreeMemoryBenchmark.benchmarkMemoryMXBeanMethod  avgt    4    0.332 ±    0.018  us/op
 * CheckingFreeMemoryBenchmark.benchmarkRuntimeMethod       avgt    4    0.075 ±    0.007  us/op
 * CheckingFreeMemoryBenchmark.benchmarkMemoryMXBeanMethod    ss    4  185.167 ± 2336.215  us/op
 * CheckingFreeMemoryBenchmark.benchmarkRuntimeMethod         ss    4    5.890 ±   46.856  us/op```

A call to rt.freeMemory() has an average penalty of 0.075 micros.